subsys/fs/ext2/ext2_disk_access.c

```diff
@@ -0,0 +1,187 @@
+} disk_data;
+
+int ext2_init_disk_access_backend(struct ext2_data *fs, const void *storage_dev, int flags)
+{
+	int rc;
+	uint32_t sector_size, sector_count;
+	const char *name = (const char *)storage_dev;
+
+	rc = disk_access_init(name);
+	if (rc < 0) {
+		LOG_ERR("FAIL: unable to find disk %s: %d\n", name, rc);
+		return rc;
+	}
+
+	rc = disk_access_ioctl(name, DISK_IOCTL_GET_SECTOR_COUNT, &sector_count);
+	if (rc < 0) {
+		LOG_ERR("Disk access (sector count) error: %d", rc);
+		return rc;
+	}
+
+	rc = disk_access_ioctl(name, DISK_IOCTL_GET_SECTOR_SIZE, &sector_size);
+	if (rc < 0) {
+		LOG_ERR("Disk access (sector size) error: %d", rc);
+		return rc;
+	}
+
+	disk_data = (struct disk_data) {
+		.name = storage_dev,
+		.sector_size = sector_size,
+		.sector_count = sector_count,
+	};
+
+	fs->backend = &disk_data;
+
+	fs->get_device_size = disk_access_device_size;
+	fs->get_write_size = disk_access_write_size;
+	fs->read_block = disk_access_read_block;
+	fs->write_block = disk_access_write_block;
+	fs->read_superblock = disk_access_read_superblock;
+	fs->sync = disk_access_sync;
+
+	return 0;
+}
+
+static int64_t disk_access_device_size(struct ext2_data *fs)
+{
+	struct disk_data *disk = fs->backend;
+
+	return disk->sector_count * disk->sector_size;
+}
+
+static int64_t disk_access_write_size(struct ext2_data *fs)
+{
+	struct disk_data *disk = fs->backend;
+
+	return disk->sector_size;
+}
+
+static void wait_for_disk(const char *dname)
+{
+	int rc = -1;
+
+	while (rc != 0) {
+		/* If we have to wait for disk then it is done inside ioctl call */
+		rc = disk_access_ioctl(dname, DISK_IOCTL_CTRL_SYNC, NULL);
+	}
+}
+
+static int disk_read(const char *disk, uint8_t *buf, uint32_t start, uint32_t num)
+{
+	int rc, loop = 0;
+
+	do {
+		wait_for_disk(disk);
+		rc = disk_access_read(disk, buf, start, num);
+		LOG_DBG("disk read: (start:%d, num:%d) (ret: %d)", start, num, rc);
+	} while ((rc == -EBUSY) && (loop++ < 16));
+	return rc;
+}
+
+static int disk_write(const char *disk, const uint8_t *buf, uint32_t start, uint32_t num)
+{
+	int rc, loop = 0;
+
+	do {
+		wait_for_disk(disk);
+		rc = disk_access_write(disk, buf, start, num);
+		LOG_DBG("disk write: (start:%d, num:%d) (ret: %d)", start, num, rc);
+	} while ((rc == -EBUSY) && (loop++ < 16));
+	return rc;
+}
+
+static int disk_prepare_range(struct disk_data *disk, uint32_t addr, uint32_t size,
+		uint32_t *s_start, uint32_t *s_count)
+{
+	/* Size must be multiple of sector_size */
+	if (size % disk->sector_size != 0) {
+		return -EINVAL;
+	}
+
+	/* Address must be aligned to sector_size */
+	if (addr & (disk->sector_size - 1)) {
+		return -EINVAL;
+	}
+
+	*s_start = CONFIG_EXT2_DISK_STARTING_SECTOR + addr / disk->sector_size;
+	*s_count = size / disk->sector_size;
+
+	LOG_DBG("addr:0x%x size:0x%x -> sector_start:%d sector_count:%d",
+			addr, size, *s_start, *s_count);
+
+	/* Cannot read or write outside the disk. */
+	if (*s_start + *s_count > disk->sector_count) {
+		return -ENOSPC;
+	}
```

Comment:
I don't think I can get rid of that. I have to check somewhere if sectors that we operate will fit in the disk otherwise we could access the sectors that are outside of used disk (they might not exist or be used to store some different data).